From c6590919fc571ed57930586672b2d418d07ef672 Mon Sep 17 00:00:00 2001
From: Zeng Heng <zengheng4@huawei.com>
Date: Tue, 29 Apr 2025 22:49:05 +0800
Subject: [PATCH 711/960] downstream: fs/resctrl: Create l2 cache monitors

Support MPAM to create *mon_L2_${cache_id}/l2c_occupancy* cache occupancy
monitors under mon_data directory. In the other hand, RDT doesn't support
L2 monitor feature.

In addition, let MPAM L2 cache msc support memory bandwidth monitor
feature. L2 cache msc not only has the MSMON_CSU monitor, but also has the
MSMON_MBWU monitor. Now provide *mon_L2_${cache_id}/mbm_core_bytes*
interface to display the value of l2c MSMON_MBWU counter.

Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/platform/mpam/mpam_devices.c | 10 +-----
 drivers/platform/mpam/mpam_resctrl.c | 50 ++++++++++++++++++++++++----
 fs/resctrl/monitor.c                 | 14 ++++++++
 fs/resctrl/rdtgroup.c                | 17 ++++++++--
 include/linux/arm_mpam.h             |  2 ++
 include/linux/resctrl_types.h        | 10 ++++++
 6 files changed, 85 insertions(+), 18 deletions(-)

diff --git a/drivers/platform/mpam/mpam_devices.c b/drivers/platform/mpam/mpam_devices.c
index 1b37bde8f33a..d6c3dcc387a7 100644
--- a/drivers/platform/mpam/mpam_devices.c
+++ b/drivers/platform/mpam/mpam_devices.c
@@ -921,17 +921,9 @@ static u64 mpam_msmon_overflow_val(struct mpam_msc_ris *ris)
 		return GENMASK_ULL(30, 0);
 }
 
-static bool resctrl_arch_would_mbm_overflow(void)
-{
-	return read_cpuid_implementor() != ARM_CPU_IMP_HISI;
-}
-
 bool resctrl_arch_would_mbm_overflow(void)
 {
-	if (is_midr_in_range_list(read_cpuid_id(), mbwu_flowrate_list))
-		return false;
-
-	return true;
+	return read_cpuid_implementor() != ARM_CPU_IMP_HISI;
 }
 
 static void __ris_msmon_read(void *arg)
diff --git a/drivers/platform/mpam/mpam_resctrl.c b/drivers/platform/mpam/mpam_resctrl.c
index adf0cb1cc37e..b14081e8c789 100644
--- a/drivers/platform/mpam/mpam_resctrl.c
+++ b/drivers/platform/mpam/mpam_resctrl.c
@@ -35,6 +35,7 @@ static bool exposed_alloc_capable;
 static bool exposed_mon_capable;
 static struct mpam_class *mbm_local_class;
 static struct mpam_class *mbm_total_class;
+static struct mpam_class *mbm_core_class;
 
 /*
  * MPAM emulates CDP by setting different PARTID in the I/D fields of MPAM1_EL1.
@@ -78,6 +79,11 @@ bool resctrl_arch_is_mbm_total_enabled(void)
 	return mbm_total_class;
 }
 
+bool resctrl_arch_is_mbm_core_enabled(void)
+{
+	return mbm_core_class;
+}
+
 bool resctrl_arch_get_cdp_enabled(enum resctrl_res_level rid)
 {
 	switch (rid) {
@@ -263,6 +269,8 @@ static void *resctrl_arch_mon_ctx_alloc_no_wait(struct rdt_resource *r,
 	case QOS_L3_OCCUP_EVENT_ID:
 	case QOS_L3_MBM_LOCAL_EVENT_ID:
 	case QOS_L3_MBM_TOTAL_EVENT_ID:
+	case QOS_L2_OCCUP_EVENT_ID:
+	case QOS_L2_MBM_CORE_EVENT_ID:
 		*ret = __mon_is_rmid_idx;
 		return ret;
 
@@ -328,10 +336,12 @@ int resctrl_arch_rmid_read(struct rdt_resource	*r, struct rdt_domain *d,
 
 	switch (eventid) {
 	case QOS_L3_OCCUP_EVENT_ID:
+	case QOS_L2_OCCUP_EVENT_ID:
 		type = mpam_feat_msmon_csu;
 		break;
 	case QOS_L3_MBM_LOCAL_EVENT_ID:
 	case QOS_L3_MBM_TOTAL_EVENT_ID:
+	case QOS_L2_MBM_CORE_EVENT_ID:
 		type = mpam_feat_msmon_mbwu;
 		break;
 	default:
@@ -480,6 +490,11 @@ bool resctrl_arch_is_llc_occupancy_enabled(void)
 	return cache_has_usable_csu(mpam_resctrl_exports[RDT_RESOURCE_L3].class);
 }
 
+bool resctrl_arch_is_l2c_occupancy_enabled(void)
+{
+	return cache_has_usable_csu(mpam_resctrl_exports[RDT_RESOURCE_L2].class);
+}
+
 static bool class_has_usable_mbwu(struct mpam_class *class)
 {
 	struct mpam_props *cprops = &class->props;
@@ -873,19 +888,23 @@ static int mpam_resctrl_resource_init(struct mpam_resctrl_res *res)
 		 * be local. If it's on the memory controller, its assumed to
 		 * be global.
 		 */
-		if (has_mbwu && class->level >= 3) {
-			mbm_local_class = class;
-			r->mon_capable = true;
+		if (has_mbwu) {
+			if (class->level == 3) {
+				mbm_local_class = class;
+				r->mon_capable = true;
+
+			} else if (class->level == 2) {
+				mbm_core_class = class;
+				r->mon_capable = true;
+			}
 		}
 
 		/*
 		 * CSU counters only make sense on a cache. The file is called
 		 * llc_occupancy, but its expected to the on the L3.
 		 */
-		if (has_csu && class->type == MPAM_CLASS_CACHE &&
-		    class->level == 3) {
+		if (has_csu && class->type == MPAM_CLASS_CACHE)
 			r->mon_capable = true;
-		}
 		break;
 
 	case RDT_RESOURCE_MBA:
@@ -1458,6 +1477,11 @@ static struct mon_evt llc_occupancy_event = {
 	.evtid		= QOS_L3_OCCUP_EVENT_ID,
 };
 
+static struct mon_evt l2c_occupancy_event = {
+	.name		= "l2c_occupancy",
+	.evtid		= QOS_L2_OCCUP_EVENT_ID,
+};
+
 static struct mon_evt mbm_total_event = {
 	.name		= "mbm_total_bytes",
 	.evtid		= QOS_L3_MBM_TOTAL_EVENT_ID,
@@ -1468,6 +1492,11 @@ static struct mon_evt mbm_local_event = {
 	.evtid		= QOS_L3_MBM_LOCAL_EVENT_ID,
 };
 
+static struct mon_evt mbm_core_event = {
+	.name		= "mbm_core_bytes",
+	.evtid		= QOS_L2_MBM_CORE_EVENT_ID,
+};
+
 /*
  * Initialize the event list for the resource.
  *
@@ -1490,6 +1519,14 @@ static void l3_mon_evt_init(struct rdt_resource *r)
 			list_add_tail(&mbm_local_event.list, &r->evt_list);
 	}
 
+	if (r->rid == RDT_RESOURCE_L2) {
+		if (resctrl_arch_is_l2c_occupancy_enabled())
+			list_add_tail(&l2c_occupancy_event.list, &r->evt_list);
+
+		if (resctrl_arch_is_mbm_core_enabled())
+			list_add_tail(&mbm_core_event.list, &r->evt_list);
+	}
+
 	if ((r->rid == RDT_RESOURCE_MBA) &&
 	     resctrl_arch_is_mbm_total_enabled())
 		list_add_tail(&mbm_total_event.list, &r->evt_list);
@@ -1498,6 +1535,7 @@ static void l3_mon_evt_init(struct rdt_resource *r)
 int resctrl_arch_mon_resource_init(void)
 {
 	l3_mon_evt_init(resctrl_arch_get_resource(RDT_RESOURCE_L3));
+	l3_mon_evt_init(resctrl_arch_get_resource(RDT_RESOURCE_L2));
 	l3_mon_evt_init(resctrl_arch_get_resource(RDT_RESOURCE_MBA));
 
 	if (resctrl_arch_is_evt_configurable(QOS_L3_MBM_TOTAL_EVENT_ID)) {
diff --git a/fs/resctrl/monitor.c b/fs/resctrl/monitor.c
index 402985f25ac6..cbb8fc878cde 100644
--- a/fs/resctrl/monitor.c
+++ b/fs/resctrl/monitor.c
@@ -574,6 +574,20 @@ static void mbm_update(struct rdt_resource *r, struct rdt_domain *d,
 		if (is_mba_sc(NULL))
 			mbm_bw_count(closid, rmid, &rr);
 
+		resctrl_arch_mon_ctx_free(rr.r, rr.evtid, rr.arch_mon_ctx);
+	}
+	if (resctrl_arch_is_mbm_core_enabled()) {
+		rr.evtid = QOS_L2_MBM_CORE_EVENT_ID;
+		rr.val = 0;
+		rr.arch_mon_ctx = resctrl_arch_mon_ctx_alloc(rr.r, rr.evtid);
+		if (IS_ERR(rr.arch_mon_ctx)) {
+			pr_warn_ratelimited("Failed to allocate monitor context: %ld",
+					    PTR_ERR(rr.arch_mon_ctx));
+			return;
+		}
+
+		__mon_event_count(closid, rmid, &rr);
+
 		resctrl_arch_mon_ctx_free(rr.r, rr.evtid, rr.arch_mon_ctx);
 	}
 }
diff --git a/fs/resctrl/rdtgroup.c b/fs/resctrl/rdtgroup.c
index cc40b178ddee..b4fb125a7509 100644
--- a/fs/resctrl/rdtgroup.c
+++ b/fs/resctrl/rdtgroup.c
@@ -113,13 +113,15 @@ void rdt_staged_configs_clear(void)
 static bool resctrl_is_mbm_enabled(void)
 {
 	return (resctrl_arch_is_mbm_total_enabled() ||
-		resctrl_arch_is_mbm_local_enabled());
+		resctrl_arch_is_mbm_local_enabled() ||
+		resctrl_arch_is_mbm_core_enabled());
 }
 
 static bool resctrl_is_mbm_event(int e)
 {
-	return (e >= QOS_L3_MBM_TOTAL_EVENT_ID &&
-		e <= QOS_L3_MBM_LOCAL_EVENT_ID);
+	return (e == QOS_L3_MBM_TOTAL_EVENT_ID ||
+		e == QOS_L3_MBM_LOCAL_EVENT_ID ||
+		e == QOS_L2_MBM_CORE_EVENT_ID);
 }
 
 /*
@@ -3842,6 +3844,15 @@ static int domain_setup_mon_state(struct rdt_resource *r, struct rdt_domain *d)
 			return -ENOMEM;
 		}
 	}
+	if (resctrl_arch_is_mbm_core_enabled()) {
+		tsize = sizeof(*d->mbm_core);
+		d->mbm_core = kcalloc(idx_limit, tsize, GFP_KERNEL);
+		if (!d->mbm_core) {
+			bitmap_free(d->rmid_busy_llc);
+			kfree(d->mbm_core);
+			return -ENOMEM;
+		}
+	}
 
 	return 0;
 }
diff --git a/include/linux/arm_mpam.h b/include/linux/arm_mpam.h
index 9f9fc7c29130..0c2a2515ff87 100644
--- a/include/linux/arm_mpam.h
+++ b/include/linux/arm_mpam.h
@@ -64,6 +64,8 @@ static inline bool resctrl_arch_event_is_free_running(enum resctrl_event_id evt)
 bool resctrl_arch_alloc_capable(void);
 bool resctrl_arch_mon_capable(void);
 bool resctrl_arch_is_llc_occupancy_enabled(void);
+bool resctrl_arch_is_l2c_occupancy_enabled(void);
+bool resctrl_arch_is_mbm_core_enabled(void);
 bool resctrl_arch_is_mbm_local_enabled(void);
 bool resctrl_arch_is_mbm_total_enabled(void);
 bool resctrl_arch_would_mbm_overflow(void);
diff --git a/include/linux/resctrl_types.h b/include/linux/resctrl_types.h
index fd1704766c29..3d56c35877ea 100644
--- a/include/linux/resctrl_types.h
+++ b/include/linux/resctrl_types.h
@@ -99,11 +99,21 @@ enum resctrl_res_level {
 /*
  * Event IDs, the values match those used to program IA32_QM_EVTSEL before
  * reading IA32_QM_CTR on RDT systems.
+ *
+ * Monitor Event IDs, representative a variety of monitoring events:
+ * QOS_L3_OCCUP_EVENT_ID:     L3 Cache Occupancy statistics event
+ * QOS_L3_MBM_TOTAL_EVENT_ID: Global Memory Bandwidth statistics event
+ * QOS_L3_MBM_LOCAL_EVENT_ID: L3 Cache Bandwidth statistics event
+ * QOS_L2_OCCUP_EVENT_ID:     L2 Cache Occupancy statistics event
+ * QOS_L2_MBM_CORE_EVENT_ID:  L2 Cache Bandwidth statistics event
  */
 enum resctrl_event_id {
 	QOS_L3_OCCUP_EVENT_ID		= 0x01,
 	QOS_L3_MBM_TOTAL_EVENT_ID	= 0x02,
 	QOS_L3_MBM_LOCAL_EVENT_ID	= 0x03,
+
+	QOS_L2_OCCUP_EVENT_ID,
+	QOS_L2_MBM_CORE_EVENT_ID,
 };
 
 #endif /* __LINUX_RESCTRL_TYPES_H */
-- 
2.43.0

